### DISCUSSION TITLE
Better Typescript support for select()!

### DISCUSSION THREAD

**Discussion by yankeeinlondon** [supabase/supabase#602] [Feature Requests]
I am using [`io-ts`](https://github.com/gcanti/io-ts) to provide strong type codec's for all tables in Postgres but in trying to provide better type support for the querying of tables I am running into issues with how types are exposed through `@supabase/supabase-js`. As a "for instance", let's say I want to provide a type-aware "select" operation for a typescript type/interface called **ISong**. Let's also assume that this table is represented in Postgres as the "songs" table.

Structurally, the operation would look like this in an untyped world:

```ts
const songs = await client.from("songs").select("artist,song").eq("artist", "Jimi Hendrix");
```

Fortunately you provide the generic <T> on the `from()` function where you _should_ be able to construct a typed selected operation but the signature of select is inheritantly weakly typed and breaks the chain. Therefore, the following typed structure:

```ts
const songs = await client.from<ISong>("songs").select("artist,song").eq("artist", "Jimi Hendrix");
```

Ideally the above would ensure that within select, only valid column names could be chosen but structurally you're asking for a single comma delimited string and this will never work. You do, however, forward the typing onward to the filter operation so _do_ get strong typing on column name and column's value in the `eq(x,y)` operation:

![image](https://user-images.githubusercontent.com/980757/105637550-04a94300-5e23-11eb-829a-435878144bad.png)

Making the select operation type-safe wouldn't be hard and I have worked around it with my wrapper by doing the following:

```ts
  return (...columns: TableProps[] | ["*"]) => {
    const cols =
      columns.length === 0 || (columns.length === 1 && columns[0]) === "*"
        ? "*"
        : columns.join(",");

    // receives a _typed_ filter operator which can 
    const filter = client.from<IModel>(tableName).select(cols);

    return filter;
  };
```

With this in place my wrapper allows me to simply define an `io-ts` type and then configure supabase like so:

```ts
const api = useSupabase(url, token).addTable(Song);

const songs = await api.table.songs.select('artist','song').eq('artist', 'Jimi Hendrix');
```

Everything in the chain -- including the table name -- is strongly typed. I got _around_ the select's break but it would be nicer if there was a strongly typed way built in. I realize this probably forces a breaking change for your API but maybe there's a way to keep the old means while offering a type-safe option?

**Comment by yankeeinlondon** [comment]
Note: it would also be very helpful to expose your interfaces -- the implementation is unnecessary -- of elements like `PostgrestFilterBuilder` on your `@supabase/supabase-js` named exports. This allows folks to be more explicit in their own type definitions rather than needing to rely on implicit inference.

**Reply by stupidawesome** [reply]
I'm playing around with this myself at the moment. [Implementation here](https://gist.github.com/stupidawesome/cb72ec4841960255e455d765114d36df)

```ts
export const bookingQuery = (id: number) => 
    query("one_time_booking")
        .eq("id", id)
        .return({
            column: ["id", "computed"],
            expand: {
                guest: "guest_profile", // short for guest_profile(*)
                // guest: {
                //    id: "guest_profile",
                //    column: ["id", "avatar_url"],
                //    expand: {
                //        ...etc
                //    }
                // }
            }
        })
        .single()
```

will produce

```gql
data: {
  id,
  computed,
  guest: {
    id,
    ...etc
  }
}
```

The "select" statement needs to come last so you can filter on props that aren't being returned.

**Comment by beeequeue** [comment]
I have two PRs open that would improve TS QoL:
https://github.com/supabase/supabase-js/pull/125
https://github.com/supabase/postgrest-js/pull/152
But they are waiting for more info regarding type generation improvements

**Reply by kiwicopple** [reply]
These are looking pretty good even as they are @BeeeQueue, but because they are breaking we will want to get as much of the storage updates in before we roll up to 2.0. I completely agree with your postgrest approach - it's much nicer to pass the definitions at init.
